### ISSUE TITLE
Configure Gradle's toolchains

### ISSUE THREAD

**Issue by dodie** [spring-io/initializr#1227]
# Steps to reproduce
- Select the following properties on https://start.spring.io/:
  - Project: Gradle Project
  - Language: Java
  - Spring Boot: 2.4.5
  - Java: 16
- Generate project

# Behavior
When I try to build the project with `./gradlew assemble` Java 16, it fails:
```
> startup failed:
  General error during semantic analysis: Unsupported class file major version 60
  
  java.lang.IllegalArgumentException: Unsupported class file major version 60
```

It seems that the generated project contains a `build.gradle` file with `sourceCompatibility = '11'`. Also, the project has Gradle version `6.8.3` which does not support Java 16 out of the box and the project does not use Gradle `toolchain`.

**Comment by wilkinsona** [comment]
start.spring.io is doing its best here to accommodate your request. Spring Boot 2.4.x only supports Gradle 6.x and Gradle 6.x does not support Java 16. GAs a result, the Java version that you have selected has been downgraded to a version that Gradle 6.8.x does support.

If you want to use Java 16 with Spring Boot 2.4.5 and Gradle, then, as you have noted, you can configure Gradle's toolchain support. We considered making start.spring.io do this automatically but decided against it as it may lead to some surprising behaviour as, for example, it can cause Gradle to download a JDK which may be blocked or prohibited for some users.

I wonder if we could/should add an entry to `HELP.md` that explains what has been downgraded and why.

**Comment by snicoll** [comment]
Isn't this a duplicate of https://github.com/spring-io/start.spring.io/issues/250 ?

**Comment by wilkinsona** [comment]
I'd forgotten about  spring-io/start.spring.io#250. Thanks for the reminder, @snicoll. I think this could well be a duplicate. 

An alternative would be for us to treat this a little bit differently. For example, we could generate the project with Gradle 6.8.3 and  `sourceCompatibility = '16'` and then describe in `HELP.md` that Gradle's Toolchain support needs to be configured to build the project. It's not a huge amount of configuration for the user to add and describing what to do without doing it by default would alleviate our concern about unexpected behaviour.

Another alternative would be to reconsider our concerns and configure the toolchain by default after all, with a note in HELP.md highlighting what has been done.

**Comment by snicoll** [comment]
Honoring the user's choice and generate a broken project with a note in HELP.md does not sound very appealing to me. If they really want to run on a version that the version of Spring Boot they've chosen is not supported, they could change back the value to 16 and follow that notes about configuring toolchain. I think this one has the advantage of being more explicit (i.e. failing only if you start using Java 12+ APIs).

**Comment by wilkinsona** [comment]
I think it could be argued that the current behaviour is also broken, albeit silently. The user's asked for something that we could configure for them, but we've given them something else instead. This is why a suggested the second alternative where we configure the toolchain. I think I'm leaning towards this approach now. It's giving the user exactly what they've asked for within the limitations of Gradle 6.x.

**Comment by snicoll** [comment]
Yup, not denying the problem, just giving feedback on the first solution and forgetting to provide some for the other one :) What you said. If configuring the toolchain is easy enough, I'd like us to explore that option and add a note in HELP.md perhaps?

**Comment by vpavic** [comment]
From today's standpoint, would you be open to configure Gradle's toolchain by default, as originally proposed in #1187?

There are numerous benefits of using Gradle toolchain on top of the case expressed in this issue. A project that configures toolchain is ensured to always build using the appropriate Java version. With the ecosystem becoming increasingly split between 3 different Java version (8, 11 & 17) this feature just becomes more attractive.

> it can cause Gradle to download a JDK which may be blocked or prohibited for some users

Isn't this basically the same as with the Gradle wrapper itself? You still use the wrapper in generated projects even though some environments might be more prohibitive about it and prefer to use their own Gradle distribution. Personally I wouldn't conflate such considerations with initializr.

**Comment by wilkinsona** [comment]
> Isn't this basically the same as with the Gradle wrapper itself?

No, I didn't think so. You can avoid using the wrapper with changing the build script by running `gradle` rather than `./gradlew`. Once a project's being configured to use a toolchain, I don't think it's possible to avoid using it without changing the build script.

**Comment by vpavic** [comment]
The way I see it:

- Gradle wrapper
  - if you don't use it you might end up using the wrong Gradle version which could fail
  - if you use it, it will download the needed Gradle version only once
- Gradle toolchains
  - if you don't use it you might end up using the wrong JDK version which could fail
  - if you use it, it will either pick it up from [the supported locations](https://docs.gradle.org/current/userguide/toolchains.html#sec:auto_detection) or download the needed JDK only once

That looks fairly analogous to me. And a far better developer experience OOTB, at least IMO.

> Once a project's being configured to use a toolchain, I don't think it's possible to avoid using it without changing the build script.

I've been using toolchains for quite some time now and in practice it has never downloaded a JDK for me, because it will pick up the appropriate JDK from my local SDKMAN installation.

Additionally, you might also take into consideration that for some time now Gradle promotes toolchains as the way of declaring Java version (see the very first snippet from the [_Building Java & JVM projects_](https://docs.gradle.org/current/userguide/building_java_projects.html) page of the user manual) and that `sourceCompatibility` and `targetCompatibility` are referred to as [_historical options for the Java compiler_](https://docs.gradle.org/current/userguide/building_java_projects.html#sec:java_cross_compilation).

**Comment by wilkinsona** [comment]
As I read it, `sourceCompatibility` and `targetCompatibility` are historical as they have been superseded by `release` not because of toolchains. In fact,  [Gradle’s own docs suggest that using release may be the easier option in some cases](https://docs.gradle.org/current/userguide/building_java_projects.html#compiling_and_testing_java_67).

In short, there’s no perfect one-size-fits-all solution here. There are numerous options here each with slightly different advantages and disadvantages, particularly when you consider that we also have to support Java 8.

**Comment by KotlinIsland** [comment]
Why not toolchains by default? I think they are awesome.

**Comment by wilkinsona** [comment]
Please see the discussion above, particularly [this comment](https://github.com/spring-io/initializr/issues/1227#issuecomment-824656869) and [the issue](https://github.com/spring-io/initializr/issues/1187) to which it links.

**Comment by wilkinsona** [comment]
That's a really nice improvement. Looks like it [landed in Gradle 8.0](https://docs.gradle.org/8.0/release-notes.html#java-toolchains-improvements). Given that it removes one of the downsides of the toolchains approach, I'm in favour of this now for cases where the Gradle version and the Java version are not compatible. Right now, we'd use a toolchain for Gradle projects that are using Java 22. When we upgrade to Gradle 8.8, we'd switch it off again.

**Comment by mhalbritter** [comment]
We talked about that in our meeting and want to make toolchains the default. If that leads to problems, we'll just use them if we need to, e.g. compiling Java 22 with Gradle 8.7.
